File: ai-service/src/services/plagiarism_service.py
"""
Plagiarism Detection Service
Uses:
  1. Winnowing algorithm — document fingerprinting (token k-grams, sliding window)
  2. AST similarity — structural code comparison (ignores variable names)
  3. No ML models or sklearn needed

Why Winnowing?
  - Creates a compact fingerprint of code by picking minimum hashes from sliding windows
  - Jaccard similarity of fingerprints detects copied sections even after renaming
  - Guaranteed to detect any copied block longer than (k + w - 1) tokens

Why AST?
  - Parses code into a tree of types/structure (ignoring identifiers & literals)
  - Two submissions that renamed all variables still show near-100% AST similarity
  - Combined with Winnowing gives robust detection
"""
import ast
import re
import hashlib
import logging
from typing import List, Dict, Optional, Set, Tuple
from dataclasses import dataclass
from collections import deque
from datetime import datetime

logger = logging.getLogger(__name__)


# ── Winnowing ─────────────────────────────────────────────────────────────────

class Winnowing:
    """
    Winnowing algorithm for code fingerprinting.
    Reference: Schleimer et al., "Winnowing: Local Algorithms for Document Fingerprinting"

    Steps:
    1. Normalize code (remove comments, lowercase, collapse whitespace)
    2. Tokenize into meaningful tokens
    3. Generate k-grams (overlapping windows of k tokens)
    4. Hash each k-gram
    5. Slide a window of size w over hashes; select the minimum in each window
    6. The set of selected hashes = fingerprint
    7. Compare via Jaccard: |A ∩ B| / |A ∪ B|
    """

    def __init__(self, k: int = 7, w: int = 5):
        """
        k: k-gram size (larger = fewer false positives, may miss short copies)
        w: window size (guarantee threshold = k + w - 1 tokens)
        """
        self.k = k
        self.w = w
        self.guarantee_threshold = k + w - 1  # Any match >= this length WILL be detected

    def fingerprint(self, code: str, language: str = "python") -> Set[int]:
        tokens = self._tokenize(code, language)
        if len(tokens) < self.k:
            return {self._hash(t) for t in tokens}

        # k-grams
        kgrams = [' '.join(tokens[i:i+self.k]) for i in range(len(tokens) - self.k + 1)]
        hashes = [self._hash(kg) for kg in kgrams]

        # Winnowing: pick min hash in each sliding window
        fingerprints: Set[int] = set()
        window: deque = deque(maxlen=self.w)
        prev_min = None

        for h in hashes:
            window.append(h)
            if len(window) == self.w:
                cur_min = min(window)
                if cur_min != prev_min:
                    fingerprints.add(cur_min)
                    prev_min = cur_min

        return fingerprints

    def similarity(self, fp1: Set[int], fp2: Set[int]) -> float:
        if not fp1 and not fp2:
            return 1.0
        if not fp1 or not fp2:
            return 0.0
        intersection = len(fp1 & fp2)
        union = len(fp1 | fp2)
        return intersection / union

    def _tokenize(self, code: str, language: str) -> List[str]:
        """Remove comments, normalize identifiers, keep structure tokens"""
        # Remove comments
        if language in ("java", "cpp", "javascript"):
            code = re.sub(r'//.*$', '', code, flags=re.MULTILINE)
            code = re.sub(r'/\*.*?\*/', '', code, flags=re.DOTALL)
        elif language == "python":
            code = re.sub(r'#.*$', '', code, flags=re.MULTILINE)
            code = re.sub(r'""".*?"""|\'\'\'.*?\'\'\'', '', code, flags=re.DOTALL)

        # Normalize string literals → STR, numbers → NUM
        code = re.sub(r'"[^"]*"|\'[^\']*\'', 'STR', code)
        code = re.sub(r'\b\d+(\.\d+)?\b', 'NUM', code)

        # Extract meaningful tokens (identifiers + operators + punctuation)
        tokens = re.findall(r'\b\w+\b|[{}()\[\];=+\-*/<>!&|]', code)

        # Normalize variable/function names to generic tokens
        keywords = {
            # Python
            'def', 'class', 'return', 'if', 'else', 'elif', 'for', 'while',
            'import', 'from', 'in', 'not', 'and', 'or', 'True', 'False', 'None',
            'try', 'except', 'finally', 'with', 'as', 'pass', 'break', 'continue',
            'lambda', 'yield', 'print', 'range', 'len', 'self',
            # Java/C++
            'int', 'long', 'void', 'bool', 'boolean', 'string', 'String', 'char',
            'public', 'private', 'static', 'new', 'null', 'true', 'false',
            'this', 'extends', 'implements', 'interface',
            # JS
            'const', 'let', 'var', 'function', 'arrow', 'async', 'await',
            'undefined', 'typeof', 'instanceof',
            # Structure tokens — keep as-is
            'STR', 'NUM',
        }
        operators_and_punctuation = set('{}()[];=+-*/<>!&|')
        normalized = []
        for t in tokens:
            if t in keywords or t in operators_and_punctuation or t in ('STR', 'NUM'):
                normalized.append(t)
            else:
                normalized.append('ID')  # Normalize user-defined names to ID

        return normalized

    def _hash(self, text: str) -> int:
        return int(hashlib.md5(text.encode()).hexdigest()[:8], 16)


# ── AST Similarity ────────────────────────────────────────────────────────────

class ASTSimilarity:
    """
    AST-based similarity for Python code.
    Builds a canonical string from the AST node types (ignoring variable names
    and literal values). Two submissions that differ only in names/values show
    near-100% similarity.
    """

    def get_ast_signature(self, code: str) -> Optional[str]:
        """Returns a string of node types from the Python AST, ignoring names/values"""
        try:
            tree = ast.parse(code)
            parts = []
            for node in ast.walk(tree):
                type_name = type(node).__name__
                # Skip pure value nodes — we want structure
                if type_name in ('Load', 'Store', 'Del', 'Add', 'Sub', 'Mult', 'Div',
                                  'Mod', 'Pow', 'LShift', 'RShift', 'BitOr', 'BitXor',
                                  'BitAnd', 'FloorDiv', 'Invert', 'Not', 'UAdd', 'USub',
                                  'Eq', 'NotEq', 'Lt', 'LtE', 'Gt', 'GtE', 'Is', 'IsNot',
                                  'In', 'NotIn', 'And', 'Or'):
                    continue
                # Skip Name/Constant values — only keep type
                parts.append(type_name)
            return ' '.join(parts)
        except Exception:
            return None

    def similarity(self, code1: str, code2: str, language: str) -> float:
        if language != 'python':
            return self._structure_similarity_regex(code1, code2, language)

        sig1 = self.get_ast_signature(code1)
        sig2 = self.get_ast_signature(code2)

        if sig1 is None or sig2 is None:
            return self._structure_similarity_regex(code1, code2, language)

        # Token-level Jaccard on AST type sequences
        set1 = set(sig1.split())
        set2 = set(sig2.split())
        if not set1 and not set2:
            return 1.0
        if not set1 or not set2:
            return 0.0
        return len(set1 & set2) / len(set1 | set2)

    def _structure_similarity_regex(self, code1: str, code2: str, language: str) -> float:
        """Fallback: compare control structure counts"""
        patterns = [
            r'\bfor\b', r'\bwhile\b', r'\bif\b', r'\belse\b',
            r'\bswitch\b', r'\breturn\b', r'\bbreak\b', r'\bcontinue\b',
        ]
        vec1 = [len(re.findall(p, code1)) for p in patterns]
        vec2 = [len(re.findall(p, code2)) for p in patterns]
        total = sum(max(a, b) for a, b in zip(vec1, vec2))
        if total == 0:
            return 0.5
        agreement = sum(min(a, b) for a, b in zip(vec1, vec2))
        return agreement / total


# ── Main Service ──────────────────────────────────────────────────────────────

@dataclass
class SimilarityResult:
    submission1_id: str
    submission2_id: str
    user1_id: str
    user2_id: str
    overall_similarity: float
    winnowing_similarity: float
    ast_similarity: float
    is_suspicious: bool


class PlagiarismService:
    """
    Orchestrates plagiarism detection using Winnowing + AST.
    No ML models needed.
    """

    def __init__(self, threshold: float = 0.75):
        self.threshold = threshold
        self.winnowing = Winnowing(k=7, w=5)
        self.ast_sim   = ASTSimilarity()

    async def check_contest(self, contest_id: str, submissions: List[Dict]) -> Dict:
        """Check all pairs in a contest. O(n²) comparisons."""
        if len(submissions) < 2:
            return {
                "contest_id": contest_id,
                "total_submissions": len(submissions),
                "suspicious_pairs": [],
                "average_similarity": 0.0,
                "checked_at": datetime.utcnow().isoformat(),
            }

        # Pre-compute fingerprints
        processed = []
        for sub in submissions:
            fp = self.winnowing.fingerprint(sub.get("code", ""), sub.get("language", "python"))
            processed.append({**sub, "fingerprint": fp})

        suspicious_pairs = []
        all_similarities = []

        for i in range(len(processed)):
            for j in range(i + 1, len(processed)):
                result = self._compare(processed[i], processed[j])
                all_similarities.append(result.overall_similarity)
                if result.is_suspicious:
                    suspicious_pairs.append(result)

        return {
            "contest_id": contest_id,
            "total_submissions": len(submissions),
            "suspicious_pairs": [
                {
                    "submission1_id":    p.submission1_id,
                    "submission2_id":    p.submission2_id,
                    "user1_id":          p.user1_id,
                    "user2_id":          p.user2_id,
                    "similarity_score":  round(p.overall_similarity, 3),
                    "winnowing_similarity": round(p.winnowing_similarity, 3),
                    "ast_similarity":    round(p.ast_similarity, 3),
                    "is_suspicious":     p.is_suspicious,
                }
                for p in suspicious_pairs
            ],
            "average_similarity": round(
                sum(all_similarities) / len(all_similarities) if all_similarities else 0.0, 3
            ),
            "checked_at": datetime.utcnow().isoformat(),
        }

    async def compare_pair(self, sub1: Dict, sub2: Dict) -> Dict:
        """Compare exactly two submissions"""
        fp1 = self.winnowing.fingerprint(sub1.get("code", ""), sub1.get("language", "python"))
        fp2 = self.winnowing.fingerprint(sub2.get("code", ""), sub2.get("language", "python"))
        p1 = {**sub1, "fingerprint": fp1}
        p2 = {**sub2, "fingerprint": fp2}
        result = self._compare(p1, p2)
        return {
            "submission1_id":       result.submission1_id,
            "submission2_id":       result.submission2_id,
            "overall_similarity":   round(result.overall_similarity, 3),
            "winnowing_similarity": round(result.winnowing_similarity, 3),
            "ast_similarity":       round(result.ast_similarity, 3),
            "is_suspicious":        result.is_suspicious,
            "threshold":            self.threshold,
        }

    def _compare(self, s1: Dict, s2: Dict) -> SimilarityResult:
        lang1 = s1.get("language", "python")
        lang2 = s2.get("language", "python")

        # Winnowing similarity (language-agnostic)
        w_sim = self.winnowing.similarity(s1["fingerprint"], s2["fingerprint"])

        # AST similarity (Python exact, others regex-based)
        if lang1 == lang2:
            a_sim = self.ast_sim.similarity(s1.get("code", ""), s2.get("code", ""), lang1)
        else:
            a_sim = 0.0  # Different languages → can't be plagiarism

        # Weighted: Winnowing 60%, AST 40%
        overall = 0.6 * w_sim + 0.4 * a_sim

        return SimilarityResult(
            submission1_id=str(s1.get("id", s1.get("submission_id", ""))),
            submission2_id=str(s2.get("id", s2.get("submission_id", ""))),
            user1_id=str(s1.get("user_id", "")),
            user2_id=str(s2.get("user_id", "")),
            overall_similarity=overall,
            winnowing_similarity=w_sim,
            ast_similarity=a_sim,
            is_suspicious=overall >= self.threshold,
        )


plagiarism_service = PlagiarismService()

File: backend/src/services/plagiarism.service.js
/**
 * Plagiarism Detection Service — runs entirely in Node.js, no external AI service needed.
 *
 * Algorithms:
 *   1. Winnowing  — document fingerprinting via k-gram hashing + sliding window min-hash
 *   2. Structural — control-flow token counting
 *
 * Combined score: 60% Winnowing + 40% Structural
 * Default threshold: 0.75
 *
 * Rules:
 *   - Only compares submissions from DIFFERENT users (same-user pairs are skipped)
 *   - Per user per problem: only the best submission is used
 *     (accepted > latest attempt) to avoid noise from failed attempts
 */

import crypto from 'crypto';
import PlagiarismReport from '../models/plagiarism.models.js';
import Submission from '../models/submission.models.js';
import ContestSubmission from '../models/postgres/ContestSubmission.models.js';
import User from '../models/user.models.js';
import logger from '../config/logger.js';
import redis from '../config/redis.config.js';

// ── Winnowing ─────────────────────────────────────────────────────────────────

function normalize(code, language) {
  let c = code || '';
  if (['java', 'cpp', 'c', 'javascript', 'typescript'].includes(language)) {
    c = c.replace(/\/\/.*$/gm, '').replace(/\/\*[\s\S]*?\*\//g, '');
  } else if (language === 'python') {
    c = c.replace(/#.*$/gm, '').replace(/"""[\s\S]*?"""|'''[\s\S]*?'''/g, '');
  }
  c = c.replace(/"[^"]*"|'[^']*'/g, 'STR');
  c = c.replace(/\b\d+(\.\d+)?\b/g, 'NUM');
  return c;
}

const KEYWORDS = new Set([
  'def','class','return','if','else','elif','for','while','import','from',
  'in','not','and','or','True','False','None','try','except','finally',
  'with','as','pass','break','continue','lambda','yield','print','range','len','self',
  'int','long','void','bool','boolean','string','String','char',
  'public','private','static','new','null','true','false',
  'this','extends','implements','interface',
  'const','let','var','function','async','await','undefined','typeof','instanceof',
  'STR','NUM',
]);

function tokenize(code, language) {
  const c = normalize(code, language);
  const raw = c.match(/\b\w+\b|[{}()[\];=+\-*/<>!&|]/g) || [];
  return raw.map(t => (KEYWORDS.has(t) || /^[{}()[\];=+\-*/<>!&|]$/.test(t)) ? t : 'ID');
}

function hashStr(s) {
  return parseInt(crypto.createHash('md5').update(s).digest('hex').slice(0, 8), 16);
}

function winnowFingerprint(code, language, k = 7, w = 5) {
  const tokens = tokenize(code, language);
  if (tokens.length < k) {
    return new Set(tokens.map(t => hashStr(t)));
  }
  const hashes = [];
  for (let i = 0; i <= tokens.length - k; i++) {
    hashes.push(hashStr(tokens.slice(i, i + k).join(' ')));
  }
  const fingerprints = new Set();
  let prevMin = null;
  const window = [];
  for (const h of hashes) {
    window.push(h);
    if (window.length > w) window.shift();
    if (window.length === w) {
      const curMin = Math.min(...window);
      if (curMin !== prevMin) {
        fingerprints.add(curMin);
        prevMin = curMin;
      }
    }
  }
  return fingerprints;
}

// ── Fingerprint cache (L4 fix) ──────────────────────────────────────────────
// A submission's code never changes after it's created, so its winnowing
// fingerprint is a pure function of immutable data — safe to cache
// indefinitely (long TTL, not a correctness concern; TTL just bounds Redis
// memory). Follows the same try/catch-and-degrade pattern as redisService.js:
// a cache failure should never break plagiarism checking, just make it
// slower for that one call.
const FINGERPRINT_CACHE_TTL = 60 * 60 * 24 * 30; // 30 days

async function getCachedFingerprint(submissionId, language) {
  try {
    const raw = await redis.get(`plagiarism:fp:${submissionId}:${language}`);
    if (!raw) return null;
    return new Set(JSON.parse(raw));
  } catch (error) {
    logger.warn(`Fingerprint cache read failed for ${submissionId}: ${error.message}`);
    return null;
  }
}

async function cacheFingerprint(submissionId, language, fingerprintSet) {
  try {
    await redis.setex(
      `plagiarism:fp:${submissionId}:${language}`,
      FINGERPRINT_CACHE_TTL,
      JSON.stringify(Array.from(fingerprintSet)),
    );
  } catch (error) {
    logger.warn(`Fingerprint cache write failed for ${submissionId}: ${error.message}`);
  }
}

// Cached wrapper around winnowFingerprint — checks Redis first, computes and
// populates on miss. Falls back to plain (uncached) computation if Redis is
// unavailable, same graceful-degrade philosophy as the rest of the codebase.
async function getFingerprint(submissionId, code, language) {
  const cached = await getCachedFingerprint(submissionId, language);
  if (cached) return cached;

  const fp = winnowFingerprint(code, language);
  await cacheFingerprint(submissionId, language, fp);
  return fp;
}

function jaccardSimilarity(setA, setB) {
  if (setA.size === 0 && setB.size === 0) return 1.0;
  if (setA.size === 0 || setB.size === 0) return 0.0;
  let intersection = 0;
  for (const h of setA) { if (setB.has(h)) intersection++; }
  const union = setA.size + setB.size - intersection;
  return intersection / union;
}

// ── Structural similarity ─────────────────────────────────────────────────────

const CONTROL_PATTERNS = [
  /\bfor\b/g, /\bwhile\b/g, /\bif\b/g, /\belse\b/g,
  /\bswitch\b/g, /\breturn\b/g, /\bbreak\b/g, /\bcontinue\b/g,
];

function structuralSimilarity(code1, code2) {
  const vec = (c) => CONTROL_PATTERNS.map(p => (c.match(p) || []).length);
  const v1 = vec(code1);
  const v2 = vec(code2);
  const total = v1.reduce((s, a, i) => s + Math.max(a, v2[i]), 0);
  if (total === 0) return 0.5;
  const agree = v1.reduce((s, a, i) => s + Math.min(a, v2[i]), 0);
  return agree / total;
}

// ── Compare a pair ────────────────────────────────────────────────────────────

function comparePair(s1, s2, threshold = 0.75) {
  const lang1 = (s1.language || 'python').toLowerCase();
  const lang2 = (s2.language || 'python').toLowerCase();

  const fp1 = s1._fp || winnowFingerprint(s1.code, lang1);
  const fp2 = s2._fp || winnowFingerprint(s2.code, lang2);

  const winnowSim = jaccardSimilarity(fp1, fp2);

  // Structural similarity only makes sense for same language.
  // For cross-language pairs we still run winnowing — identical logic
  // (e.g. same C++ submitted by two users as different "languages") will
  // still score 1.0 on winnowing alone.
  const sameLanguage = lang1 === lang2;
  const structSim    = sameLanguage ? structuralSimilarity(s1.code, s2.code) : 0.0;

  // If cross-language: weight 100% winnowing so identical code is still caught.
  // If same language: 60% winnowing + 40% structural (original weighting).
  const overallSim = sameLanguage
    ? 0.6 * winnowSim + 0.4 * structSim
    : winnowSim;

  return { winnowSim, structSim, overallSim, isSuspicious: overallSim >= threshold };
}

// ── Main Service ──────────────────────────────────────────────────────────────

class PlagiarismService {
  constructor() {
    this.threshold = 0.75;
  }

  async checkContest(contestId, checkedBy = null) {
    try {
      // Step 1 — load ALL contest submissions from PostgreSQL
      const contestSubs = await ContestSubmission.findAll({
        where: { contest_id: contestId },
        attributes: ['submission_id', 'user_id', 'problem_id', 'language', 'status', 'submitted_at'],
        order: [['submitted_at', 'ASC']],
      });

      if (contestSubs.length < 2) {
        return {
          totalSubmissions: contestSubs.length,
          suspiciousPairs:  [],
          averageSimilarity: 0,
          message: 'Not enough submissions to run plagiarism check (need at least 2)',
        };
      }

      // Step 2 — pick the BEST submission per (user, problem)
      //   Priority: accepted > latest attempt
      //   This eliminates noise from failed attempts and prevents
      //   a single user's own submissions from being compared against each other
      const bestMap = new Map(); // key: `${user_id}:${problem_id}` → best row
      for (const cs of contestSubs) {
        const key = `${cs.user_id}:${cs.problem_id}`;
        const existing = bestMap.get(key);
        if (!existing) {
          bestMap.set(key, cs);
        } else {
          // Prefer accepted over any other status
          const csAccepted = cs.status === 'accepted';
          const exAccepted = existing.status === 'accepted';
          if (csAccepted && !exAccepted) {
            bestMap.set(key, cs); // take the accepted one
          } else if (csAccepted === exAccepted) {
            bestMap.set(key, cs); // both same status — take the later one
          }
          // if existing is accepted and cs is not — keep existing
        }
      }

      const bestSubs = Array.from(bestMap.values());

      // Step 3 — load code from MongoDB
      const submissionIds = bestSubs.map(s => s.submission_id).filter(id => {
        // Only include valid MongoDB ObjectId strings (24 hex chars)
        return id && /^[a-f\d]{24}$/i.test(id);
      });

      if (submissionIds.length < 2) {
        logger.warn(`Plagiarism check: only ${submissionIds.length} valid MongoDB submission IDs found for contest ${contestId}`);
        return {
          totalSubmissions: bestSubs.length,
          suspiciousPairs:  [],
          averageSimilarity: 0,
          message: 'Could not find enough valid submission references in database',
        };
      }

      const mongoSubs = await Submission.find({ _id: { $in: submissionIds } })
        .select('_id code language user')
        .populate('user', 'username')
        .lean();

      const subMap = Object.fromEntries(mongoSubs.map(s => [s._id.toString(), s]));

      // Step 4 — merge metadata + code, get fingerprints (cached where possible)
      const submissions = (await Promise.all(
        bestSubs.map(async (cs) => {
          const mongo = subMap[cs.submission_id];
          if (!mongo?.code) return null;
          const lang = (cs.language || mongo.language || 'python').toLowerCase();
          return {
            submission_id: cs.submission_id,
            user_id:       cs.user_id,
            problem_id:    cs.problem_id,
            code:          mongo.code,
            language:      lang,
            username:      mongo.user?.username || cs.user_id,
            _fp:           await getFingerprint(cs.submission_id, mongo.code, lang),
          };
        })
      )).filter(Boolean);

      logger.info(`Plagiarism check contest ${contestId}: ${submissions.length} best submissions from ${new Set(submissions.map(s => s.user_id)).size} users`);

      if (submissions.length < 2) {
        return {
          totalSubmissions: contestSubs.length,
          suspiciousPairs:  [],
          averageSimilarity: 0,
          message: 'Could not load code for enough submissions',
        };
      }

      // Step 5 — O(n²) pair comparison
      //   ONLY compare submissions from DIFFERENT users for the SAME problem
      //   Comparing cross-problem is meaningless (different problems have different solutions)
      const suspiciousPairs = [];
      const allSimilarities = [];

      for (let i = 0; i < submissions.length; i++) {
        for (let j = i + 1; j < submissions.length; j++) {
          const s1 = submissions[i];
          const s2 = submissions[j];

          // Skip same-user pairs — a user cannot plagiarise themselves
          if (s1.user_id === s2.user_id) continue;

          // Only compare submissions for the same problem
          if (s1.problem_id !== s2.problem_id) continue;

          // NOTE: cross-language pairs are NOT skipped — comparePair() handles
          // them gracefully (winnowing-only, no structural weight).
          // This catches identical code submitted under different language labels.

          const { winnowSim, structSim, overallSim, isSuspicious } = comparePair(s1, s2, this.threshold);
          allSimilarities.push(overallSim);

          if (isSuspicious) {
            suspiciousPairs.push({
              submission1:          s1.submission_id,
              submission2:          s2.submission_id,
              user1:                s1.user_id,
              user2:                s2.user_id,
              similarityScore:      Math.round(overallSim * 1000) / 1000,
              tokenSimilarity:      Math.round(winnowSim  * 1000) / 1000,
              astSimilarity:        Math.round(structSim  * 1000) / 1000,
              structuralSimilarity: Math.round(structSim  * 1000) / 1000,
              isSuspicious:         true,
              verdict:              'pending',
            });
          }
        }
      }

      const averageSimilarity = allSimilarities.length
        ? allSimilarities.reduce((a, b) => a + b, 0) / allSimilarities.length
        : 0;

      // Step 6 — save report, preserving verdicts already reviewed by admin.
      // Without this, every re-run wipes confirmed plagiarism decisions.
      const existingReport = await PlagiarismReport.findOne({ contest: contestId }).lean();
      const existingVerdictMap = new Map();
      if (existingReport?.suspiciousPairs) {
        for (const ep of existingReport.suspiciousPairs) {
          if (ep.verdict && ep.verdict !== 'pending') {
            const key = [ep.submission1?.toString(), ep.submission2?.toString()].sort().join('|');
            existingVerdictMap.set(key, {
              verdict:     ep.verdict,
              reviewed:    ep.reviewed,
              reviewedBy:  ep.reviewedBy,
              reviewNotes: ep.reviewNotes,
            });
          }
        }
      }

      // Merge preserved verdicts into freshly computed pairs
      for (const pair of suspiciousPairs) {
        const key = [pair.submission1?.toString(), pair.submission2?.toString()].sort().join('|');
        const saved = existingVerdictMap.get(key);
        if (saved) {
          pair.verdict     = saved.verdict;
          pair.reviewed    = saved.reviewed;
          pair.reviewedBy  = saved.reviewedBy;
          pair.reviewNotes = saved.reviewNotes;
        }
      }

      await PlagiarismReport.deleteOne({ contest: contestId });
      const report = await PlagiarismReport.create({
        contest:           contestId,
        totalSubmissions:  submissions.length,
        checkedBy,
        checkedAt:         new Date(),
        status:            'completed',
        averageSimilarity: Math.round(averageSimilarity * 1000) / 1000,
        threshold:         this.threshold,
        suspiciousPairs,
        metadata: {
          algorithmsUsed: ['winnowing', 'structural'],
          version:        '2.2-local',
        },
      });

      logger.info(`Plagiarism check done for contest ${contestId}: ${suspiciousPairs.length} suspicious pair(s) among ${submissions.length} best submissions`);
      return report;

    } catch (error) {
      logger.error('Plagiarism check failed:', error.message);
      throw new Error(`Plagiarism check failed: ${error.message}`);
    }
  }

  async autoCheckOnContestEnd(contestId) {
    try {
      logger.info(`Auto-running plagiarism check for contest ${contestId}`);
      const report = await this.checkContest(contestId, null);
      const suspicious = report.suspiciousPairs?.length || 0;
      logger.info(`Auto plagiarism check done: ${suspicious} suspicious pair(s) for contest ${contestId}`);
      return report;
    } catch (err) {
      logger.error(`Auto plagiarism check failed for contest ${contestId}:`, err.message);
    }
  }

  async getReport(contestId) {
    // Use .lean() — populate() silently fails because user1/user2 are stored as
    // plain strings (MongoDB ObjectId hex), not Mongoose ObjectId references.
    // We manually enrich them instead.
    const report = await PlagiarismReport.findOne({ contest: contestId })
      .sort({ checkedAt: -1 })
      .lean();

    if (!report) {
      throw new Error('No plagiarism report found for this contest');
    }

    // Collect all unique user IDs across all suspicious pairs
    const pairs = report.suspiciousPairs || [];
    const rawIds = pairs.flatMap(p => [
      p.user1?.toString(),
      p.user2?.toString(),
    ]).filter(id => id && /^[a-f\d]{24}$/i.test(id));
    const uniqueIds = [...new Set(rawIds)];

    // Fetch usernames from MongoDB in one query
    let userMap = {};
    if (uniqueIds.length) {
      const users = await User.find({ _id: { $in: uniqueIds } })
        .select('username email profile.avatar')
        .lean();
      users.forEach(u => { userMap[u._id.toString()] = u; });
    }

    // Enrich each pair's user1/user2 with actual user data
    report.suspiciousPairs = pairs.map(p => {
      const u1id = p.user1?.toString();
      const u2id = p.user2?.toString();
      return {
        ...p,
        user1: userMap[u1id] || { _id: u1id, username: `User_${(u1id || '').slice(-6)}`, email: null },
        user2: userMap[u2id] || { _id: u2id, username: `User_${(u2id || '').slice(-6)}`, email: null },
      };
    });

    return report;
  }

  async compareTwo(submission1Id, submission2Id) {
    try {
      const [sub1, sub2] = await Promise.all([
        Submission.findById(submission1Id).select('code language user').lean(),
        Submission.findById(submission2Id).select('code language user').lean(),
      ]);
      if (!sub1 || !sub2) throw new Error('One or both submissions not found');

      const lang1 = (sub1.language || 'python').toLowerCase();
      const lang2 = (sub2.language || 'python').toLowerCase();
      const { winnowSim, structSim, overallSim, isSuspicious } = comparePair(
        { code: sub1.code, language: lang1 },
        { code: sub2.code, language: lang2 },
        this.threshold
      );

      return {
        submission1_id:       submission1Id,
        submission2_id:       submission2Id,
        overall_similarity:   Math.round(overallSim * 1000) / 1000,
        winnowing_similarity: Math.round(winnowSim  * 1000) / 1000,
        ast_similarity:       Math.round(structSim  * 1000) / 1000,
        is_suspicious:        isSuspicious,
        threshold:            this.threshold,
      };
    } catch (error) {
      logger.error('Comparison failed:', error.message);
      throw new Error('Failed to compare submissions');
    }
  }

  async reviewPair(contestId, submission1Id, submission2Id, verdict, notes, reviewedBy) {
    const report = await PlagiarismReport.findOne({ contest: contestId });
    if (!report) throw new Error('Report not found');

    const pair = report.suspiciousPairs.find(
      p =>
        (p.submission1?.toString() === submission1Id && p.submission2?.toString() === submission2Id) ||
        (p.submission1?.toString() === submission2Id && p.submission2?.toString() === submission1Id)
    );
    if (!pair) throw new Error('Pair not found in report');

    pair.reviewed    = true;
    pair.reviewedBy  = reviewedBy;
    pair.verdict     = verdict;
    pair.reviewNotes = notes;

    await report.save();
    return report;
  }
}

export default new PlagiarismService();